File: SampleApp/android/app/src/main/java/com/sampleapp/BlueshiftBridge.java
package com.sampleapp;

import androidx.annotation.NonNull;

import com.blueshift.Blueshift;
import com.blueshift.BlueshiftAppPreferences;
import com.blueshift.BlueshiftLogger;
import com.blueshift.model.UserInfo;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;

public class BlueshiftBridge extends ReactContextBaseJavaModule {
    private static final String TAG = "BlueshiftBridge";
    private ReactApplicationContext applicationContext;

    public BlueshiftBridge(ReactApplicationContext reactContext) {
        super(reactContext);
        this.applicationContext = reactContext;
        BlueshiftLogger.setLogLevel(BlueshiftLogger.VERBOSE);
    }

    @NonNull
    @Override
    public String getName() {
        return "BlueshiftBridge";
    }


    @ReactMethod
    void setUserEmailId(String email) {
        UserInfo.getInstance(applicationContext).setEmail(email);
        UserInfo.getInstance(applicationContext).save(applicationContext);
    }

    @ReactMethod
    void setUserCustomerId(String customerId) {
        UserInfo.getInstance(applicationContext).setRetailerCustomerId(customerId);
        UserInfo.getInstance(applicationContext).save(applicationContext);
    }

    @ReactMethod
    void identifyWithDetails(ReadableMap map) {
        String email = UserInfo.getInstance(applicationContext).getEmail();
        Blueshift.getInstance(applicationContext).identifyUserByEmail(email, map.toHashMap(), false);
    }

    @ReactMethod
    void trackCustomEvent(String eventName, ReadableMap map, boolean canBatch) {
        Blueshift.getInstance(applicationContext).trackEvent(eventName, map.toHashMap(), canBatch);
    }

    @ReactMethod
    void trackScreenView(String screenName, ReadableMap map, boolean canBatch) {
        Blueshift.getInstance(applicationContext).trackScreenView(screenName, canBatch);
    }

    @ReactMethod
    void removeUserInfo() {
        BlueshiftLogger.d(TAG, "Method not available");
    }

    @ReactMethod
    void registerForRemoteNotification() {
        BlueshiftLogger.d(TAG, "Method not available");
    }

    @ReactMethod
    void setEnablePush(boolean enablePush) {
        BlueshiftAppPreferences.getInstance(applicationContext).setEnablePush(enablePush);
        BlueshiftAppPreferences.getInstance(applicationContext).save(applicationContext);
    }

    @ReactMethod
    void registerForInApp(String screenName) {
        BlueshiftLogger.d(TAG, "Method not available");
    }

    @ReactMethod
    void unregisterForInApp() {
        BlueshiftLogger.d(TAG, "Method not available");
    }

    @ReactMethod
    void fetchInAppNotification() {
        Blueshift.getInstance(applicationContext).fetchInAppMessages(null);
    }

    @ReactMethod
    void displayInAppNotification() {
        Blueshift.getInstance(applicationContext).displayInAppMessages();
    }
}

File: SampleApp/ios/BlueshiftBridge.swift
//
//  BlueshiftBridge.swift
//  SampleApp
//
//  Created by Ketan Shikhare on 20/11/20.
//

import UIKit
import Foundation

@objc(BlueshiftBridge)
class BlueshiftBridge: NSObject {
  @objc static func requiresMainQueueSetup() -> Bool {return true}
  let screenViewed = "screen_viewed"
  
  // MARK: Track Events
  @objc func identifyWithDetails(_ details: Dictionary<AnyHashable, Any>) {
    BlueShift.sharedInstance()?.identifyUser(withDetails: details, canBatchThisEvent: false)
  }
  
  @objc func identifyWithEmail(_ email: String, details: Dictionary<AnyHashable, Any>) {
    BlueShift.sharedInstance()?.identifyUser(withEmail: email, andDetails: details, canBatchThisEvent: false)
  }
  
  @objc func trackCustomEvent(_ eventName: String, details: Dictionary<AnyHashable, Any> ,canBatchThisEvent: Bool) {
      BlueShift.sharedInstance()?.trackEvent(forEventName: eventName, andParameters: details, canBatchThisEvent: canBatchThisEvent)
  }
  
  @objc func trackScreenView(_ screenName: String, details: Dictionary<AnyHashable, Any> ,canBatchThisEvent: Bool) {
      var params = details
      params[screenViewed] = screenName
      BlueShift.sharedInstance()?.trackEvent(forEventName: kEventPageLoad, andParameters: params, canBatchThisEvent: canBatchThisEvent)
  }

  // MARK: Set user info
  @objc func setUserEmailId(_ emailId: String) {
    BlueShiftUserInfo.sharedInstance()?.email = emailId
    BlueShiftUserInfo.sharedInstance()?.save()
  }
  
  @objc func setUserCustomerId(_ customerId: String) {
    BlueShiftUserInfo.sharedInstance()?.retailerCustomerID = customerId
    BlueShiftUserInfo.sharedInstance()?.save()
  }
  
  @objc func removeUserInfo() {
    BlueShiftUserInfo.removeCurrentUserInfo()
  }
  
  // MARK: Push notifications
  @objc func registerForRemoteNotification() {
    BlueShift.sharedInstance()?.appDelegate.registerForNotification()
  }
  
  @objc func setEnablePush(_ isEnabled: Bool) {
    BlueShiftAppData.current()?.enablePush = isEnabled
  }
  
  // MARK: Set location for tracking
  @objc func setCurrentLocation(_ lat: Double, long:Double) {
    let location = CLLocation(latitude: lat, longitude: long)
    BlueShiftDeviceData.current()?.currentLocation = location
  }

  // MARK: Inapp notifications
  @objc func registerForInApp(_ displayPage: String) {
    DispatchQueue.main.async {
      BlueShift.sharedInstance()?.registerFor(inAppMessage: displayPage)
    }
  }
  
  @objc func unregisterForInApp() {
    BlueShift.sharedInstance()?.unregisterForInAppMessage()
  }

  @objc func fetchInAppNotification() {
    BlueShift.sharedInstance()?.fetchInAppNotification(fromAPI: (() -> Void)? {}, failure: ((Error?) -> Void)?{_ in
    })
  }
  
  @objc func displayInAppNotification() {
    DispatchQueue.main.async {
      BlueShift.sharedInstance()?.displayInAppNotification()
    }
  }
    
}
